feat(NX-3251): add drawer element

araujobarret opened the pull request:
Related also to https://github.com/artsy/volt-v2/pull/424 and [NX-3251]

This PR creates a drawer element following some specs we used in upcoming features that the NX team is working on.

The idea here is based on the concept of having a provider with a hook linked to toggle the Drawer on/off. Since we don't have any specs somethings in this PR we can make them flexible once we have the official specs for the Drawer, like the flex of the Overlay and Drawer Content.

Desktop

https://github.com/artsy/palette/assets/15792853/d0e0cd80-f970-4c2d-8061-3d9fbc68964e

Mobile

https://github.com/artsy/palette/assets/15792853/8adb3f67-ca71-44f8-9414-73d2e1ae89d7

araujobarret commented on packages/palette/src/elements/Drawer/Drawer.tsx:
```diff
@@ -0,0 +1,98 @@
+    return css`
+      transform: translateX(${props.anchor === "left" ? -100 : 100}%);
+      flex-direction: ${props.anchor === "left" ? "row" : "row-reverse"};
+      transition-duration: ${props.isOpen ? "300ms" : "6000ms"};
```
I'm not an animation ninja, but if I use a duration like 300ms when it's not open it goes stupidity fast, any insights why are more than welcome

dzucconi replied:
I _really_ feel strongly that designers should be designing this.

araujobarret commented on packages/palette/src/elements/Drawer/Drawer.tsx:
```diff
@@ -0,0 +1,98 @@
+export interface DrawerProps {
+  anchor?: "left" | "right"
+  flexContent?: number
+  flexOverlay?: number
```
any recommendation for this proportion case? Or we can hardcode the flex value proportion until we have a DS official size?

dzucconi replied:
I'd imagine the size is entirely dependent on the content? So that might change depending on use case. We're going to be using this for the filters for instance, which will be pretty specific. Just setting defaults that are over-writable with `BoxProps` is the way to go. I just read these props and have no idea what they mean really. We could have specific `contentProps` and `overlayProps` slots that each accept `BoxProps` is also fine here.

dzucconi reviewed (commented):
An important consideration here is accessibility. Let me see if I can dig up the aria pattern for this kind of thing. At a minimum we need to be managing focus: trapping it within this layer when it's open, and yielding it back to the trigger when it's closed. See the `ModalBase` for instance.

araujobarret commented on packages/palette/src/elements/Drawer/DrawerProvider.tsx:
```diff
@@ -0,0 +1,29 @@
+  const close = useCallback(() => setIsOpen(false), [])
+
+  return (
+    <DrawerContext.Provider value={{ isOpen, toggle, close }}>
```
the idea here is more like the toaster, to enable the drawer to be called from multiple components without many declarations, but I can change it to be like the Modal style. I'll poke around to see how it works without it 👍

dzucconi commented:
Hm, [no 'official' guidance](https://github.com/w3c/aria-practices/issues/1590), though this is basically a modal with a different shape. [Chakra calls out the same things](https://chakra-ui.com/docs/components/drawer#accessibility) regarding focus management at least.

araujobarret commented:
I did the requested changes/ideas [here](https://github.com/artsy/palette/pull/1295/commits/07e49023e9ffb09a19c736001ecda7d9bdf49ab1), summary of what I did:
- User react-focus-on wrapping the drawer content
- Removed Context API(using similar approach of ModalBase now)
- simplified the animation using animation API instead of transitioning
- no out animation(we're not caring about it too much right now in Palette)

Result:

https://github.com/artsy/palette/assets/15792853/cef2fc1b-87fc-4097-b397-bc832380f0e2

If everything is fine here, should I drop a message in DS channel asking for a Designer to help craft the animation, maybe merging here, asking it to them, and having a follow-up PR with the more accurate animation?

damassi commented on packages/palette/src/elements/Drawer/Drawer.tsx:
```diff
@@ -0,0 +1,131 @@
+  height: 100%;
+  align-items: center;
+  justify-content: center;
+`
```
The animation CSS / props are spread over a pretty wide surface area in this file and i think it can be simplified considerably. Naively thinking this through, I could see the below semi-pseudocode being all that's needed:

```tsx
const Container = styled(Box)`
  opacity: 0; 
  right: -500px;
  
  .open {
    opacity: 1;
    right: 0;
    transition: ease-out; 
    transition-duration: 300ms; 
  }
`

Something like this; if state is to open, just add an open class. Or to avoid classNames, pass the open prop into the component and use the styled-components `css` helper. But yah, i'd try to get the LOC count in here down to the bare essentials.

damassi replied:
Yah `transform: translateX(...)`, something on the gpu. `right` gets the idea across 😄

dzucconi commented on packages/palette/src/elements/Drawer/Drawer.tsx:
```diff
@@ -0,0 +1,131 @@
+  )
+}
+
+const DEFAULT_DRAWER_Z_INDEX = 9999
```
This should just be like 1

araujobarret replied:
I thought we'd follow a similar pattern as we have in the [BaseModal](https://github.com/artsy/palette/blob/main/packages/palette/src/elements/Modal/ModalBase.tsx#L53-L56), but I changed in https://github.com/artsy/palette/pull/1295/commits/657df35bf582513603b7d209058a22b6a3093a62

dzucconi replied:
Good observation that that comment probably needs more context and less snark! At the time I was migrating legacy modals and that was what the value was at the time.

araujobarret replied:
Ah, I should have guessed something came up during past migrations, but no worries from my side, default 1 seems to be reasonable 👍

dzucconi commented on packages/palette/src/elements/Drawer/Drawer.tsx:
```diff
@@ -0,0 +1,131 @@
+          overflowX="hidden"
+          overflowY="scroll"
+          anchor={anchor}
+          zIndex={zIndex}
```
Doesn't need z-index here

araujobarret commented on packages/palette/src/elements/Drawer/Drawer.tsx:
```diff
@@ -0,0 +1,131 @@
+          height="100%"
+          width={["100%", "auto"]}
+          overflowX="hidden"
+          overflowY="scroll"
```
done in https://github.com/artsy/palette/pull/1295/commits/657df35bf582513603b7d209058a22b6a3093a62

damassi commented on packages/palette/src/elements/Drawer/Drawer.tsx:
```diff
@@ -0,0 +1,105 @@
+
+  &.open {
+    transform: translateX(0);
+  }
```
Why do we need to transform both container and content? Can we just animate one single thing?

araujobarret replied:
That's a good catch! I was supposed to remove the `transform` from the Container 👍

araujobarret replied:
changed in https://github.com/artsy/palette/pull/1295/commits/d34d6525758407f332ba51f43422e0ad35784ce0

araujobarret commented:
@dzucconi using `react-focus-on` I noticed we have by default `autoFocus={true}`, do we really want it always like this or do you think it's better to expose it throughout the Drawer API?

dzucconi commented:
Generally speaking we do want to focus the first item in the drawer (modal) once it's opened. Hard to imagine a case where we wouldn't want this?

araujobarret commented:
I share the same feeling, was more to check if I'm missing something in the API, thanks 👍

dzucconi commented on packages/palette/src/elements/Drawer/Drawer.tsx:
```diff
@@ -0,0 +1,104 @@
+          overflowY="scroll"
+          anchor={anchor}
+          zIndex={zIndex}
+          className={openClass}
```
Question: why target this using a `className`? You can use `${Content} {`

araujobarret replied:
No reasoning, guilty of being away from pre/post-processed CSS too much time that forgot to use many features it has 👍 
changed on https://github.com/artsy/palette/pull/1295/commits/842e521adcd786313b04c02df8c367c06fe699e5 (I also had to change the order of declarations to comply with linter rules)

dzucconi commented on packages/palette/src/elements/Drawer/Drawer.tsx:
```diff
@@ -0,0 +1,103 @@
+  const openClass = open ? "open" : ""
+
+  return (
+    <Container zIndex={zIndex} anchor={anchor} className={openClass}>
```
No need to pass the classname any longer.

araujobarret replied:
done in https://github.com/artsy/palette/pull/1295/commits/0a669fce9351066ec0b313301a2805dea7a9d8fb

artsyit commented:
<!-- GITHUB_RELEASE COMMENT: released -->
:rocket: PR was released in [`@artsy/palette-charts@32.2.0`](https://github.com/artsy/palette/releases/tag/%40artsy/palette-charts%4032.2.0), [`@artsy/palette@33.2.0`](https://github.com/artsy/palette/releases/tag/%40artsy/palette%4033.2.0) :rocket:

artsyit commented:
<!-- GITHUB_RELEASE COMMENT: released -->
:rocket: PR was released in [`@artsy/palette-charts@36.0.0`](https://github.com/artsy/palette/releases/tag/%40artsy/palette-charts%4036.0.0), [`@artsy/palette@37.0.0`](https://github.com/artsy/palette/releases/tag/%40artsy/palette%4037.0.0) :rocket:
